relayer/relayer.go

```diff
@@ -408,19 +410,52 @@ func (r *Relayer) waitForTransactionAndRetryIfNeeded(ctx context.Context, ethCli
 		_, err := r.EVMClient.WaitForTransaction(ctx, ethClient, newTx, r.RetryTimeout)
 		if err != nil {
 			if stderrors.Is(err, context.DeadlineExceeded) {
-				newGasPrice, err := ethClient.SuggestGasPrice(ctx)
-				if err != nil {
-					return err
-				}
-				if newGasPrice.Uint64() <= newTx.GasPrice().Uint64() {
-					// no need to resend the transaction if the suggested gas price is lower than the original one
-					continue
+				var rawTx *coregethtypes.Transaction
+				if tx.GasPrice() != nil {
+					rawTx, err = createSpeededUpLegacyTransaction(ctx, ethClient, newTx)
+					if err != nil {
+						return err
+					}
+					if rawTx.GasPrice().Cmp(newTx.GasPrice()) <= 0 {
+						// no need to resend the transaction if the suggested gas price is lower than the original one
+						continue
+					}
+				} else if tx.GasTipCap() != nil && tx.GasFeeCap() != nil {
+					rawTx, err = createSpeededUpDynamicTransaction(ctx, ethClient, newTx)
+					if err != nil {
+						return err
+					}
+					if rawTx.GasFeeCap().Cmp(newTx.GasFeeCap()) <= 0 {
+						// no need to resend the transaction if the suggested gas price is lower than the original one
+						continue
+					}
+				} else {
+					// Only query for basefee if gasPrice not specified
+					if head, errHead := ethClient.HeaderByNumber(ctx, nil); errHead != nil {
+						return errHead
+					} else if head.BaseFee != nil {
+						rawTx, err = createSpeededUpDynamicTransaction(ctx, ethClient, newTx)
+						if err != nil {
+							return err
+						}
+						if rawTx.GasFeeCap().Cmp(newTx.GasFeeCap()) <= 0 {
+							// no need to resend the transaction if the suggested gas price is lower than the original one
+							continue
+						}
+					} else {
+						// Chain is not London ready -> use legacy transaction
+						rawTx, err = createSpeededUpLegacyTransaction(ctx, ethClient, newTx)
+						if err != nil {
+							return err
+						}
+						if rawTx.GasPrice().Cmp(newTx.GasPrice()) <= 0 {
+							// no need to resend the transaction if the suggested gas price is lower than the original one
+							continue
+						}
+					}
 				}
```

Comment:
I don't think its necessarily worth changing, but it might be a good exercise to try and refactor this. Submitting a transaction is always really frustrating and some elaborate logic is unavoidable, I also haven't tried to so it might not be possible. However, my gut instinct is if there are this many if, if else, else, and continue statements there usually a way to rewrite it.

Reply:
That's a good idea, but I think that's the best we could do because we have multiple chain types so we need to speed up the transaction according to its type. This implementation is inspired from the geth implementation.